src/ixdat/tools.py

```diff
@@ -71,3 +80,159 @@ def list_is_close(list_one, list_two):
+    if hard_deprecation_release is not None:
+        message += (
+            "* It will become hard deprecated, raising exceptions rather than "
+            f"issuing warnings, from version {hard_deprecation_release}\n"
+        )
+    else:
+        message += (
+            "* It will continue to be soft deprecated, issuing warnings, for the "
+            "foreseeable future\n"
+        )
+
+    # Add information of potential removal
+    if remove_release is not None:
+        message += f"* It is planned for complete removal in version {remove_release}\n"
+
+    message += (
+        "\n"
+        "See instructions below on how to update your code to avoid this message:\n"
+        f"{update_message}"
+    )
+    return message
+
+
+def deprecate(
+    last_supported_release,
+    update_message,
+    hard_deprecation_release=None,
+    remove_release=None,
+    kwarg_name=None,
+):
+    """Mark a function, method or class for deprecation
+
+    The deprecator supports soft and hard deprecation, which will either issue warnings
+    or raise exceptions, as well as providing information about an update path and the
+    potential time the functionality will be completely removed.
+
+    Args:
+        last_supported_release (str): The name of the last version when the deprecated
+            functionality was fully supported e.g. "1.2.3"
+        update_message (str): A message to the user with instructions on how to upgrade
+            to avoid the deprecated functionality
+
+    Keyword Args:
+        hard_deprecation_release (str): The release with which the deprecation will raise
+            exceptions rather than issue warnings e.g. "1.4.0". The default None means
+            that this deprecation will remain soft for the foreseeable future.
+        remove_release (str): The release with which the deprecated functionality is
+            planned for deletion
+        kwarg_name (str): If given, is the name of the keyword argument which is
+            deprecated. If not given it is assumed that the entire callable is
+            deprecated.
+
+    Examples:
+
+    Used to deprecate a class::
+     @deprecate("1.2.3", "Please use `MyNewClass` instead", "1.4.0", "2.0.0")
+     class MyClass:
+         ...
+
+    Used to deprecate a method or an argument in a method::
+        class MyClass:
+
+            @deprecate("1.2.3", "Please use `mynewmethod` instead", "1.4.0", "2.0.0")
+            def mymethod(self):
+                ...
+
+            @classmethod
+            @deprecate(
+                "1.2.3",
+                "Please use `kwargb` instead",
+                "1.4.0",
+                "2.0.0",
+                kwarg_name="kwarga"
+            )
+            def mymethod(cls, arg, kwarga=None, kwargb=None):
+                ...
+
+    .. note::
+       In the example above, when this decorator is applied to a class method or
+       static method, it must be applied as the first decorator (closest to the def).
```

Comment:
I don't know. I kind of thought that I was already over engineering when I extended it to static and class methods. I will check.

Reply:
It does. But as with the other "special" method types, it needs to be applied closest to the def.

This works:
```python
from ixdat.tools import deprecate

class A:

    def __init__(self):
        self._a = 8

    @property
    @deprecate("1.2.3", "Just no")
    def a(self):
        return self._a

    @a.setter
    @deprecate("1.2.3", "Just no")
    def a(self, value):
        self._a = value


a = A()
print(a.a)
a.a = 10
print(a.a)
```

I will update the docs to show that this works.

Reply:
Awesome. that will be useful. 
ixdat v0.2+ uses `Measurement.add_calibration(my_cal)` instead of `Measurement.calibration = my_cal` , so I can make a dummy property and deprecate its setter to inform the user :)